### ISSUE TITLE
[wxStyledTextCtrl] Upgrade wxSTC version to latest

### ISSUE THREAD

**Issue by eranif** [wxWidgets/wxWidgets#24971]
The current version of `wxStyledTextCtrl` is based on Scintilla `5.0.0`.
Is it possible to upgrade it to the latest? (version 5.0.0 is roughly **30** versions behind the latest)

Thanks!

**Comment by vadz** [comment]
AFAIK later versions of STC require C++17 while we still support C++11. We could have 2 versions of STC, with people using C++17 getting the latest and fallback to the current one for the others, but this risks being complicated.

We could also just say that while the rest of wx only requires C++11, wxSTC requires C++17, which I think shouldn't be a huge problem in practice, but this will make things messier...

**Comment by eranif** [comment]
What about new API in `stc.h`? Do I need to run some kind of python scripts or new methods are added manually?
Also, I did notice that wx are using their own branches for Lexilla / Scintilla - we should probably worry about it later, right?

I am guessing that making the upgrade to 5.1 should probably work for 5.4?

**Comment by eranif** [comment]
By your describe of the task, I hope I will have time to do this (I have zero knowledge on sci internals)...

**Comment by rudolfwalter** [comment]
I also think this would be nice to have.

I did some work on this over the Christmas break. I don't intend to finish it, but if any of you think you'll find my changes useful as a starting point, you can find them here:
 - https://github.com/rudolfwalter/scintilla-wx/tree/upgrade-5-5-3
     - this has the most changes, and as far as I remember, it is complete
 - https://github.com/rudolfwalter/lexilla-wx/tree/upgrade-5-4-1
     - this has some small changes; I saw that Maarten made equivalent ones since
 - https://github.com/rudolfwalter/wxWidgets/tree/scintilla-5-5-3
     - this has fixes for being able to compile with the two above, but there are missing pieces

Nothing has been tested.


### Additional notes:

**Scintilla:**

I merged release 5.5.3 into the wx branch, and I made a lot of fixes to get the code to compile without errors (on Windows with VS 2022 in default mode (C++14) and on Linux with Clang 10.0 in C++11 mode).

It might be worth noting that I reverted some of the existing compatility changes, particularly the ones changing std::string_view into explicit char*/int pairs, as I find that unsustainable. Instead, I made a largely API-compatible partial reimplementation of std::string_view as Sci::string_view in Compat.h. Same for std::optional.

I tried keeping a relatively clean commit history here, in case it helps.

**wxSTC:**

I made most of the code compile with the new wxscintilla (including gen_iface.py changes).

More work needs to be done here. I left a few comments in the code marked with "TODO(rwg)" where I believe things are incomplete.

From what I can tell, be biggest missing piece is that there are new functions in Scintilla::Internal::Surface that need to be implemented in SurfaceImpl and SurfaceD2D (in PlatWX.cpp). I didn't do this because I suspect it will be more efficient for someone who is already familiar with this code to do it. Until this is done, the stc project fails to build.

*PS:* @vadz Anything I say or contribute from this account I do as a private individual, unless I explicitly state otherwise.

**Comment by LegalizeAdulthood** [comment]
> The current version of `wxStyledTextCtrl` is based on Scintilla `5.0.0`.

I'm confused.  When I checkout tag v3.2.6 (the latest release), I get scintilla 3.7.2 mentioned in `src/stc/scintilla/README.txt` and no mention of scintilla in `.gitmodules`.  I'm consuming wxWidgets via vcpkg.

**Comment by vadz** [comment]
@LegalizeAdulthood Scintilla 5.0 is in master. We're not going to update Scintilla version in 3.2 (well, not until someone goes to superhuman efforts to do it, but this seems unlikely).

@rudolfwalter Thanks, I think having polyfills for `std::string_view` and `std::optional` is indeed a good idea (and would allow using them in the main wx code too, not just here), but I don't know the surface-related code well either, unfortunately.

**Comment by LegalizeAdulthood** [comment]
@vadz Ah, I see.  So what are we losing by being on the old 3.7.2 version of scintilla?  I couldn't find docs for this version online; browsing around it appears the big difference is that we're on the old integrated lexers version and not the version that separated out lexilla.  I assume we're also missing some "big features" as we're two major versions behind, but it's hard for me to identify what's missing to tell if I "need" it or not `:)`.

**Comment by rudolfwalter** [comment]
@LegalizeAdulthood https://www.scintilla.org/ScintillaHistory.html

**Comment by LegalizeAdulthood** [comment]
Yeah, I saw that, but it's a huge changelog and doesn't indicate what is a major feature and what is some sort of tiny tweak.  Oh well.

**Comment by LegalizeAdulthood** [comment]
BTW, I did a [video on writing a lexer for a custom language](https://www.youtube.com/watch?v=M07iGUA07rY) with the current version.

**Comment by mtangoo** [comment]
Hi,
Did anyone do anything beyond this conversation?
Also, did I read correctly that the master include new version of scintilla/Lexilla where it supports more lexers than 3.x, albeit not the latest?

I have been out for quiet very long as far as wx is concerned and would like to upgrade our little Editor to support Kotlin syntax

**Comment by MaartenBent** [comment]
Lexilla is almost up-to-date (5.4.4, latest available is 5.4.5), but it does not have a Kotlin lexer.
Scintilla is still outdated and at 5.0.

**Comment by mtangoo** [comment]
Thanks for the comment. I assume you mean current master have lexilla 5.4.4. That's sweet!

What is a major thing I miss with sticking to Scintilla 5.0? 

As of Kotlin lexer, i alsonfound the same, but then it seems notepad3 have it implemented and their license is MIT. So I will try to integrate that and then try to make PR to wxwidgets and then to upstream

**Comment by MaartenBent** [comment]
Indeed, in master.
wx3.2 tries to be compatible with old compilers without c++11, so it is stuck at 3.7.2.

The changelog of Scintilla is here: https://scintilla.org/ScintillaHistory.html
I don't use wxStyledTextCtrl myself, so I don't know if there are any major features/fixes we're missing out on.

**Comment by mtangoo** [comment]
Thanks. I Will upgrade the app to master in coming weeks and see if there is any major missing features. The fact that wx uses new Scintilla/Lexilla is a huge win already!

I'll update master and try to get the Kotlin lexer working!
